### DISCUSSION TITLE
[Feature request] Attachments as links to external URLs

### DISCUSSION THREAD

**Discussion by jpbaril** [sysadminsmedia/homebox#266] [Ideas]
### What is the problem you are trying to solve with this feature?

I use [Paperelss-NGX](https://github.com/paperless-ngx/paperless-ngx) to store receipts/manuals/etc. and so would prefer to keep all my documents in there. Yet, I would want to reference those external documents in Homebox items.

### What is the solution you are proposing?

My first thought would be that when adding an "attachment" Homebox would allow to choose between uploading a file to Homebox (as it is right now) or simply link to an external entity/URL.

That way Homebox could continue to reference these external entities as per the usual attachment types (i.e. photos, receipt, warranty, etc.) and so show/use them just like those uploaded to Homebox.

### What alternatives have you considered?

I could create each time a custom field for receipts+manuals+etc. and input the external URLs since custom fields allow URL links, but that would not be "clean" and Homebox would not understand "system-wise" what those linked entities are (i.e. receipts/manuals/etc). And having to create a custom field for each file that needs to be linked to is cubersome and can become messy in my opinion. 

### Additional context

My proposition is really just a general idea / concept, there is maybe a better way to achieve the same goal.

BTW, it's a copy-pasting of a feature request I created on original repo: https://github.com/hay-kot/homebox/issues/704
Also, another user requested the exact same thing a few weeks after: https://github.com/hay-kot/homebox/issues/835 

Thank you

### Contributions

- [X] I have searched through existing issues and feature requests to see if my idea has already been proposed.
- [ ] If this feature is accepted, I would be willing to help implement and maintain this feature.
- [ ] If this feature is accepted, I'm willing to sponsor the development of this feature.

**Comment by tankerkiller125** [comment]
I've converted this into an issue for further discussion and investigation about how this might best be addressed and handled. And for further community feedback.

**Comment by kusold** [comment]
Homebox's handling of attachments is giving me some hesitation. Currently (0.15.2) stores attachments in `/data/${groupUUID}/documents/${attachmentUUID}.${fileExtension}`. While using UUIDs for the attachments on the filesystem prevents potential naming collisions, they no longer are human readable. Additionally, attachments aren't included in the export CSV which makes rebuilding a Homebox instance nigh impossible if the database is lost.

The proposed solution of allowing attachments be URLs works well for PDF data. I also run Paperless-NGX and how it handles exporting documents is superb. If https://github.com/sysadminsmedia/homebox/pull/332 adds support for [uploading directly to Paperless](https://docs.paperless-ngx.com/api/#file-uploads), that would solve some of my concerns.

There would still need to be a way to Export attachment references so that users can recover from database corruption, or have a way to move their data to a new instance.

Potentially Related Issues:
* https://github.com/sysadminsmedia/homebox/issues/224
* https://github.com/sysadminsmedia/homebox/issues/288
* https://github.com/sysadminsmedia/homebox/pull/332

**Reply by tankerkiller125** [reply]
While the filename is not human readable, the original name is stored in the database (although not really used except in the UI at the moment).

Moving data to a new instance is already possible by simply copying the database file to the new location (unless you mean something like importing the data to a different shared instance or something?)

As for corruption, or loss of the db file, properly working backups would resolve that pretty quickly.

**Reply by kusold** [reply]
I just wanted to come back and say that you were correct that my request doesn't belong in Homebox's code base. Thanks to the API that homebox provides, I was able to implement what I wanted.

https://github.com/kusold/homebox-export

It's pretty basic at the moment, but it currently is able to pull back all of my attachments with their original name. Furthermore, by using the API it should also work with future storage providers. Eventually I might also start exporting the items themselves since the CSV export is a bit limited with the fields it exports.

**Comment by samturner3** [comment]
Related: https://github.com/sysadminsmedia/homebox/discussions/544

**Comment by samuelkdavis** [comment]
Samsung Food has a nice feature that I believe could be relevant here. In Samsung Food you can paste in a recipe, and it prompts to ask if you'd like to import the recipe or link to it.

Import = take a copy of the details and never update
Link = periodically scan the source

**Comment by Aristocrat-jt** [comment]
Jumping in here, I don't know if it's related. I had links in another app where it nicely would view PDFs in HomeBox (assuming I was logged into HomeBox for a proper token). But with the latest version it now just downloads the PDF. It would be nice if it was reverted to viewing in the browser. I believe it was the Open in New Tab button. Now it seems Open in New Tab and Download are the same thing.

**Comment by szaiser** [comment]
I’d like to add to the discussion an enhancement proposition inspired by [Triliums Webview capability](https://github.com/TriliumNext/Trilium/issues/2515)/ [Web View.md](https://github.com/TriliumNext/Trilium/blob/eeeecb39887e23b347298a22fe1ea21f3dcb1e9f/docs/User%20Guide/User%20Guide/Note%20Types/Web%20View.md) to natively integrate attachments as links to external URLs in homebox

### Proposed User Flow

1. **Drag a URL** into the attachments area of a Homebox item.
2. Homebox **detects it’s a URL** and treats it like an external attachment.

#### Expected Behavior

* Clicking the link opens the URL in the **web browser** .
* Additionally, the external URL is **shown in a webview** either **permanently** or only **on mouse hover**.

**Comment by flybrys** [comment]
Just adding my 2c. I think it would be super handy if you could set up an API connection to paperless-ngx, and reference the attachment by document ID. Homebox could then possibly retrieve the document on request each time it is viewed

**Comment by s-martin** [comment]
I would like to see such a feature as well.

My use case is I store the receipts etc in paperless-ngx and I would like to link to them.

A simple URL is totally fine, accessing paperless via API would be superb.

**Comment by meremortals70** [comment]
I too would like to be able to reference a manual, invoice etc as a URL from another system. As above and there seems to be growing community who uses paperless-ngx and would like to link directly to it. However, I can see use cases beyond paperless-ngx. It is more about having a System of Truth for documents that other apps draw from.

**Comment by elyashivhazan** [comment]
@tankerkiller125 Hi 👋🏼
Could it be done by adding a new filed called `Remote Attachments` and the old attachments will be renamed to `Local Attachments`

That way no need to add code for migration nor its braking/removing locally stored files
it will also allow for setting remote attachments using csv and also exporting

and `Remote Attachments` will only accepts HTTP URL

**Comment by szaiser** [comment]
Feature implemented in [PR 1481](https://github.com/sysadminsmedia/homebox/pull/1481). Please support by reviewing and putting comments.
